pageserver/src/tenant/mgr.rs

```diff
@@ -864,34 +876,185 @@ pub(crate) enum GetTenantError {
+    tenant_id: TenantId,
+    mut timeout: Duration,
+    cancel: &CancellationToken,
+) -> Result<Arc<Tenant>, GetActiveTenantError> {
+    enum WaitFor {
+        Barrier(utils::completion::Barrier),
+        Tenant(Arc<Tenant>),
+    }
+
+    let wait_for = {
+        let locked = TENANTS.read().unwrap();
+        let peek_slot =
+            tenant_map_peek_slot(&locked, &tenant_id, true).map_err(GetTenantError::MapState)?;
+        match peek_slot {
+            Some(TenantSlot::Attached(tenant)) => {
+                match tenant.current_state() {
+                    TenantState::Active => {
+                        // Fast path: we don't need to do any async waiting.
+                        return Ok(tenant.clone());
+                    }
+                    _ => WaitFor::Tenant(tenant.clone()),
+                }
+            }
+            Some(TenantSlot::Secondary) => {
+                return Err(GetActiveTenantError::NotFound(GetTenantError::NotActive(
+                    tenant_id,
+                )))
+            }
+            Some(TenantSlot::InProgress(barrier)) => WaitFor::Barrier(barrier.clone()),
+            None => {
+                return Err(GetActiveTenantError::NotFound(GetTenantError::NotFound(
+                    tenant_id,
+                )))
+            }
+        }
+    };
+
+    let tenant = match wait_for {
+        WaitFor::Barrier(barrier) => {
+            tracing::debug!("Waiting for tenant InProgress state to pass...");
+            let wait_start = Instant::now();
+            timeout_cancellable(timeout, barrier.wait(), cancel)
+                .await
+                .map_err(|e| match e {
+                    TimeoutCancellableError::Timeout => {
+                        GetActiveTenantError::WaitForActiveTimeout {
+                            latest_state: None,
+                            wait_time: wait_start.elapsed(),
+                        }
+                    }
+                    TimeoutCancellableError::Cancelled => GetActiveTenantError::Cancelled,
+                })?;
+            let wait_duration = Instant::now().duration_since(wait_start);
+            timeout -= wait_duration;
+            {
+                let locked = TENANTS.read().unwrap();
+                let peek_slot = tenant_map_peek_slot(&locked, &tenant_id, true)
+                    .map_err(GetTenantError::MapState)?;
+                match peek_slot {
+                    Some(TenantSlot::Attached(tenant)) => tenant.clone(),
+                    _ => {
+                        return Err(GetActiveTenantError::NotFound(GetTenantError::NotActive(
+                            tenant_id,
+                        )))
+                    }
+                }
+            }
+        }
+        WaitFor::Tenant(tenant) => tenant,
+    };
+
+    tracing::debug!("Waiting for tenant to enter active state...");
+    match timeout_cancellable(timeout, tenant.wait_to_become_active(), cancel).await {
```

Comment:
Couldn't it become `InProgress` again since we last checked? 

The dirty fix for this would be to turn this function into a loop.

That doesn't ensure fairness, though.

Reply:
This function returns an Arc<Tenant>, so it's always the case that the returned tenant's slot might have gone into an InProgress state by the time the receiver gets it.